src/shared/copy.c

```diff
@@ -1170,8 +1170,24 @@ static int fd_copy_tree_generic(
 
         r = fd_copy_leaf(df, from, st, dt, to, override_uid, override_gid, copy_flags, hardlink_context, display_path, progress_bytes, userdata);
         /* We just tried to copy a leaf node of the tree. If it failed because the node already exists *and* the COPY_REPLACE flag has been provided, we should unlink the node and re-copy. */
-        if (r == -EEXIST && (copy_flags & COPY_REPLACE)) {
-                /* This codepath is us trying to address an error to copy, if the unlink fails, lets just return the original error. */
+        if (r == -EEXIST && (copy_flags & (COPY_REPLACE|COPY_REPLACE_UPDATED))) {
+                /* This codepath is us trying to address an error to copy, if we fail, let's just return the original error. */
+
+                if (copy_flags & COPY_REPLACE_UPDATED) {
+                        nsec_t orig_ts = NSEC_INFINITY, replace_ts = NSEC_INFINITY;
+                        struct stat buf;
+
+                        replace_ts = timespec_load_nsec(&st->st_mtim);
+                        if (fstatat(dt, to, &buf, AT_SYMLINK_NOFOLLOW) >= 0)
+                                orig_ts = timespec_load_nsec(&buf.st_mtim);
+
+                        if (orig_ts == NSEC_INFINITY || replace_ts == NSEC_INFINITY)
+                                return r; /* If we couldn't get both modification times, play it safe and don't overwrite. */
+
+                        if (orig_ts >= replace_ts)
```

Comment:
hmm, why check for newer? why not check for different?

I mean, I think we should not expect that if a home dir is portable between systems that all systems have correct clocks.

I think it would be more robust to just check for different timestamps here, rather than newer/older. 

Or in other words, use something like stat_inode_unmodified().

Reply:
We have to check that the file is newer, not just different. Otherwise we'd only be able to sync changes to files one way.

Right now the relevant part of the sync algorithm looks something like this: Copy REPLACE_UPDATED sys -> home, then home -> sys. If we switch REPLACE_UPDATED to REPLACE_DIFFERENT, that's actually just the same thing as REPLACE. In other words, we always overwrite a file. So if you have /bulk/avatar and ~bulk/avatar, replacing the ~bulk/avatar will never do anything because the first copy step will immediately replace the ~bulk/avatar with /bulk/avatar (which has a different, but older, timestamp)

> I think we should not expect that if a home dir is portable between systems that all systems have correct clocks.

Well then that's unfortunate and someone who's using homed on systems with invalid clocks will just lose the files in their bulk directory occasionally :shrug: